File: step_14_bst/14.2_practice/IB114_sorted_array_to_balanced_bst/Solution.java
import java.util.*;

/**
 * Problem: Sorted Array to Balanced BST
 * Difficulty: EASY | XP: 10
 * Source: InterviewBit
 *
 * Convert a sorted array to a height-balanced BST.
 * Mid element becomes root; recurse on left and right halves.
 *
 * @author DSA_Nova
 */

class BSTNode {
    int val;
    BSTNode left, right;
    BSTNode(int val) { this.val = val; }
}

// ============================================================
// APPROACH 1: BRUTE FORCE
// Naive recursive mid-point selection
// Time: O(N)  |  Space: O(log N) stack
// ============================================================
class BruteForce {
    static BSTNode sortedArrayToBST(int[] nums, int left, int right) {
        if (left > right) return null;
        int mid = (left + right) / 2;
        BSTNode node = new BSTNode(nums[mid]);
        node.left = sortedArrayToBST(nums, left, mid - 1);
        node.right = sortedArrayToBST(nums, mid + 1, right);
        return node;
    }

    static BSTNode convert(int[] nums) {
        return sortedArrayToBST(nums, 0, nums.length - 1);
    }
}

// ============================================================
// APPROACH 2: OPTIMAL
// Same recursive approach but uses (left + right + 1) / 2 for right-biased mid
// to match LeetCode's expected output when even elements exist.
// Time: O(N)  |  Space: O(log N)
// ============================================================
class Optimal {
    static BSTNode helper(int[] nums, int left, int right) {
        if (left > right) return null;
        // Right-biased mid for even-length arrays
        int mid = left + (right - left + 1) / 2;
        BSTNode node = new BSTNode(nums[mid]);
        node.left = helper(nums, left, mid - 1);
        node.right = helper(nums, mid + 1, right);
        return node;
    }

    static BSTNode convert(int[] nums) {
        return helper(nums, 0, nums.length - 1);
    }
}

// ============================================================
// APPROACH 3: BEST
// Iterative using explicit stack — (node, left, right) triples
// Time: O(N)  |  Space: O(N) for the result tree; O(log N) stack
// ============================================================
class Best {
    static BSTNode convert(int[] nums) {
        if (nums.length == 0) return null;
        // Stack stores: [BSTNode, left, right, isLeft (1) or isRight (0), parent]
        // Use a simpler approach: store pending work
        int n = nums.length;
        int mid = n / 2;
        BSTNode root = new BSTNode(nums[mid]);

        // Queue of (node, arrayLeft, arrayRight)
        Deque<int[]> indexStack = new ArrayDeque<>();
        Deque<BSTNode> nodeStack = new ArrayDeque<>();
        Deque<Boolean> isLeftStack = new ArrayDeque<>();
        Deque<BSTNode> parentStack = new ArrayDeque<>();

        // Push right subtask
        if (mid + 1 <= n - 1) {
            indexStack.push(new int[]{mid + 1, n - 1});
            nodeStack.push(root);
            isLeftStack.push(false);
        }
        // Push left subtask
        if (0 <= mid - 1) {
            indexStack.push(new int[]{0, mid - 1});
            nodeStack.push(root);
            isLeftStack.push(true);
        }

        while (!indexStack.isEmpty()) {
            int[] range = indexStack.pop();
            BSTNode parent = nodeStack.pop();
            boolean isLeft = isLeftStack.pop();

            int l = range[0], r = range[1];
            int m = l + (r - l) / 2;
            BSTNode newNode = new BSTNode(nums[m]);

            if (isLeft) parent.left = newNode;
            else parent.right = newNode;

            if (m + 1 <= r) {
                indexStack.push(new int[]{m + 1, r});
                nodeStack.push(newNode);
                isLeftStack.push(false);
            }
            if (l <= m - 1) {
                indexStack.push(new int[]{l, m - 1});
                nodeStack.push(newNode);
                isLeftStack.push(true);
            }
        }
        return root;
    }
}

public class Solution {
    static List<Integer> inorder(BSTNode root) {
        List<Integer> res = new ArrayList<>();
        Deque<BSTNode> stack = new ArrayDeque<>();
        BSTNode cur = root;
        while (cur != null || !stack.isEmpty()) {
            while (cur != null) { stack.push(cur); cur = cur.left; }
            cur = stack.pop();
            res.add(cur.val);
            cur = cur.right;
        }
        return res;
    }

    static int height(BSTNode root) {
        if (root == null) return 0;
        return 1 + Math.max(height(root.left), height(root.right));
    }

    public static void main(String[] args) {
        System.out.println("=== Sorted Array to Balanced BST ===");

        int[] nums = {-10, -3, 0, 5, 9};
        BSTNode r1 = BruteForce.convert(nums);
        BSTNode r2 = Optimal.convert(nums);
        BSTNode r3 = Best.convert(nums);

        System.out.println("BruteForce inorder: " + inorder(r1) + " height=" + height(r1));
        System.out.println("Optimal    inorder: " + inorder(r2) + " height=" + height(r2));
        System.out.println("Best       inorder: " + inorder(r3) + " height=" + height(r3));

        int[] single = {1};
        System.out.println("Single element: " + inorder(BruteForce.convert(single)));

        int[] even = {1, 2, 3, 4};
        BSTNode re = BruteForce.convert(even);
        System.out.println("Even array inorder: " + inorder(re) + " height=" + height(re));
    }
}

File: step_14_bst/14.2_practice/IB114_sorted_array_to_balanced_bst/solution.py
"""
Problem: Sorted Array to Balanced BST
Difficulty: EASY | XP: 10
Source: InterviewBit

Convert a sorted array to a height-balanced BST.
Pick middle element as root; recurse on left and right halves.
"""

from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# ============================================================
# APPROACH 1: BRUTE FORCE
# Recursive mid-point: always pick left-biased mid
# Time: O(N)  |  Space: O(log N) recursion stack
# ============================================================
def brute_force(nums: List[int]) -> Optional[TreeNode]:
    def helper(left, right):
        if left > right:
            return None
        mid = (left + right) // 2
        node = TreeNode(nums[mid])
        node.left = helper(left, mid - 1)
        node.right = helper(mid + 1, right)
        return node

    return helper(0, len(nums) - 1)


# ============================================================
# APPROACH 2: OPTIMAL
# Right-biased mid to match expected output for even-length arrays
# Time: O(N)  |  Space: O(log N)
# ============================================================
def optimal(nums: List[int]) -> Optional[TreeNode]:
    def helper(left, right):
        if left > right:
            return None
        mid = left + (right - left + 1) // 2  # right-biased
        node = TreeNode(nums[mid])
        node.left = helper(left, mid - 1)
        node.right = helper(mid + 1, right)
        return node

    return helper(0, len(nums) - 1)


# ============================================================
# APPROACH 3: BEST
# Iterative using explicit stack of (node, left, right, is_left_child)
# Time: O(N)  |  Space: O(log N) stack
# ============================================================
def best(nums: List[int]) -> Optional[TreeNode]:
    if not nums:
        return None
    n = len(nums)
    mid = n // 2
    root = TreeNode(nums[mid])
    # stack entries: (parent_node, left_idx, right_idx, is_left)
    stack = []
    if mid + 1 <= n - 1:
        stack.append((root, mid + 1, n - 1, False))
    if 0 <= mid - 1:
        stack.append((root, 0, mid - 1, True))

    while stack:
        parent, l, r, is_left = stack.pop()
        m = l + (r - l) // 2
        new_node = TreeNode(nums[m])
        if is_left:
            parent.left = new_node
        else:
            parent.right = new_node
        if m + 1 <= r:
            stack.append((new_node, m + 1, r, False))
        if l <= m - 1:
            stack.append((new_node, l, m - 1, True))

    return root


def inorder(root):
    res = []
    stack = []
    cur = root
    while cur or stack:
        while cur:
            stack.append(cur)
            cur = cur.left
        cur = stack.pop()
        res.append(cur.val)
        cur = cur.right
    return res


def height(root):
    if not root:
        return 0
    return 1 + max(height(root.left), height(root.right))


if __name__ == "__main__":
    print("=== Sorted Array to Balanced BST ===")

    nums = [-10, -3, 0, 5, 9]
    r1 = brute_force(nums)
    r2 = optimal(nums)
    r3 = best(nums)
    print(f"BruteForce inorder: {inorder(r1)}, height={height(r1)}")  # sorted, height<=3
    print(f"Optimal    inorder: {inorder(r2)}, height={height(r2)}")
    print(f"Best       inorder: {inorder(r3)}, height={height(r3)}")

    single = [1]
    print(f"Single: {inorder(brute_force(single))}")  # [1]

    even = [1, 2, 3, 4]
    print(f"Even array inorder: {inorder(brute_force(even))}, height={height(brute_force(even))}")
